src/rowbinary/de.rs

```diff
@@ -1,94 +1,194 @@
-use std::{convert::TryFrom, mem, str};
-
 use crate::error::{Error, Result};
+use crate::rowbinary::utils::{ensure_size, get_unsigned_leb128};
+use crate::rowbinary::validation::SerdeType;
+use crate::rowbinary::validation::{DataTypeValidator, ValidateDataType};
 use bytes::Buf;
+use clickhouse_rowbinary::data_types::{Column, DataTypeHint};
+use serde::de::MapAccess;
 use serde::{
     de::{DeserializeSeed, Deserializer, EnumAccess, SeqAccess, VariantAccess, Visitor},
     Deserialize,
 };
+use std::{convert::TryFrom, mem, str};
 
 /// Deserializes a value from `input` with a row encoded in `RowBinary`.
 ///
 /// It accepts _a reference to_ a byte slice because it somehow leads to a more
 /// performant generated code than `(&[u8]) -> Result<(T, usize)>` and even
 /// `(&[u8], &mut Option<T>) -> Result<usize>`.
 pub(crate) fn deserialize_from<'data, T: Deserialize<'data>>(input: &mut &'data [u8]) -> Result<T> {
-    let mut deserializer = RowBinaryDeserializer { input };
+    // println!("deserialize_from call");
+
+    let mut deserializer = RowBinaryDeserializer {
+        input,
+        validator: (),
+    };
+    T::deserialize(&mut deserializer)
+}
+
+/// Similar to [`deserialize_from`], but expects a slice of [`Column`] objects
+/// parsed from the beginning of `RowBinaryWithNamesAndTypes` data stream.
+/// After the header, the rows format is the same as `RowBinary`.
+pub(crate) fn deserialize_from_and_validate<'data, 'cursor, T: Deserialize<'data>>(
+    input: &mut &'data [u8],
+    columns: &'cursor [Column],
+) -> Result<T> {
+    let mut deserializer = RowBinaryDeserializer {
+        input,
+        validator: DataTypeValidator::new(columns),
+    };
     T::deserialize(&mut deserializer)
 }
 
-/// A deserializer for the RowBinary format.
+/// A deserializer for the RowBinary(WithNamesAndTypes) format.
 ///
 /// See https://clickhouse.com/docs/en/interfaces/formats#rowbinary for details.
-struct RowBinaryDeserializer<'cursor, 'data> {
-    input: &'cursor mut &'data [u8],
+pub(crate) struct RowBinaryDeserializer<'cursor, 'data, Validator = ()>
+where
+    Validator: ValidateDataType,
+{
+    pub(crate) validator: Validator,
+    pub(crate) input: &'cursor mut &'data [u8],
 }
 
-impl<'data> RowBinaryDeserializer<'_, 'data> {
-    fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
+impl<'data, Validator> RowBinaryDeserializer<'_, 'data, Validator>
+where
+    Validator: ValidateDataType,
+{
+    pub(crate) fn read_vec(&mut self, size: usize) -> Result<Vec<u8>> {
         Ok(self.read_slice(size)?.to_vec())
     }
 
-    fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
+    pub(crate) fn read_slice(&mut self, size: usize) -> Result<&'data [u8]> {
         ensure_size(&mut self.input, size)?;
         let slice = &self.input[..size];
         self.input.advance(size);
         Ok(slice)
     }
 
-    fn read_size(&mut self) -> Result<usize> {
+    pub(crate) fn read_size(&mut self) -> Result<usize> {
```

Comment:
Need to revert it back; this is an artifact from the previous implementation attempt, where I wanted to leverage `deserialize_any` instead.